onnxruntime/core/framework/kernel_registry_manager.cc

```diff
@@ -14,63 +14,105 @@ Status KernelRegistryManager::CreateKernel(const onnxruntime::Node& node,
                                            const IExecutionProvider& execution_provider,
                                            const SessionState& session_state,
                                            /*out*/ std::unique_ptr<OpKernel>& op_kernel) const {
-  std::lock_guard<OrtMutex> lock(lock_);
-  if (kernel_registries_.empty()) {
-    return Status(ONNXRUNTIME, FAIL, "Kernel not found.");
+  const std::string& ptype = node.GetExecutionProviderType();
+  if (ptype.empty()) {
+    return Status(ONNXRUNTIME, FAIL, "The node is not placed on any Execution Provider, therefore, can't find a suitable kernel for it");
   }
-
   Status status;
-  for (auto& registry : kernel_registries_) {
-    status = registry->CreateKernel(node,
-                                    execution_provider,
-                                    session_state.GetInitializedTensors(),
-                                    session_state.GetMLValueNameIdxMap(),
-                                    session_state.GetFuncMgr(),
-                                    op_kernel);
+  {
+    std::lock_guard<OrtMutex> lock(lock_);
+    for (auto& registry : kernel_registries_) {
+      status = registry->CreateKernel(node,
+                                      execution_provider,
+                                      session_state.GetInitializedTensors(),
+                                      session_state.GetMLValueNameIdxMap(),
+                                      session_state.GetFuncMgr(),
+                                      op_kernel);
+      if (status.IsOK()) {
+        return status;
+      }
+    }
+  }
+
+  KernelRegistry* p = nullptr;
+  {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto iter = provider_type_to_registry_.find(ptype);
+    if (iter != provider_type_to_registry_.end()) p = iter->second.get();
+  }
+  if (p != nullptr) {
+    status = p->CreateKernel(node,
+                             execution_provider,
+                             session_state.GetInitializedTensors(),
+                             session_state.GetMLValueNameIdxMap(),
+                             session_state.GetFuncMgr(),
+                             op_kernel);
     if (status.IsOK()) {
       return status;
     }
   }
```

Comment:
A comment explaining why this second phase of kernel creation is needed would be helpful. Also, we call CreateKernel under the lock on line 25, but outside of the lock on line 44. Why the differenece?